packages/core/src/browser/tree/tree-selection.ts

```diff
@@ -67,45 +100,109 @@ export namespace ISelectableTreeNode {
 @injectable()
 export class TreeSelectionService implements ITreeSelectionService {
 
-    protected _selectedNode: ISelectableTreeNode | undefined;
-    protected readonly onSelectionChangedEmitter = new Emitter<ISelectableTreeNode | undefined>();
-
-    constructor( @inject(ITree) protected readonly tree: ITree) { }
+    @inject(ITree)
+    protected readonly tree: ITree;
+    protected readonly _selectedNodes: ISelectableTreeNode[] = [];
+    protected readonly onSelectionChangedEmitter = new Emitter<TreeSelection>();
 
     dispose() {
         this.onSelectionChangedEmitter.dispose();
     }
 
-    get selectedNode(): ISelectableTreeNode | undefined {
-        return this._selectedNode;
+    get selectedNodes(): TreeSelection {
+        return this._selectedNodes.slice();
     }
 
-    get onSelectionChanged(): Event<ISelectableTreeNode | undefined> {
+    get onSelectionChanged(): Event<TreeSelection> {
```

Comment:
why is it not simply an array of selectable tree nodes?

Reply:
It will remove unnecessary array copying by clients

Reply:
This is my goal:

```
export interface StructuredSelection<T> extends ReadonlyArray<Readonly<T>> { }

export interface TreeSelection extends StructuredSelection<ISelectableTreeNode> { }
```

So, I would like to keep the `TreeSelection` as a type, and the `TreeSelectionService` will make sure to call `slice()` before firing the event or when calling the getter for the selected items. What's your take on that?